receiver/prometheus/metrics_receiver.go

```diff
@@ -15,36 +15,90 @@
 package prometheus
 
 import (
+	"context"
 	"errors"
+	"sync"
 	"time"
 
+	agentmetricspb "github.com/census-instrumentation/opencensus-proto/gen-go/agent/metrics/v1"
+	"github.com/census-instrumentation/opencensus-service/data"
 	"github.com/census-instrumentation/opencensus-service/receiver"
+	"github.com/orijtech/promreceiver"
+	"github.com/prometheus/prometheus/config"
 )
 
-// Receiver is the type used to handle metrics from OpenCensus exporters.
-type Receiver struct {
-	metricSink         receiver.MetricsReceiverSink
-	metricBufferPeriod time.Duration
-	metricBufferCount  int
+type Configuration struct {
+	ScrapeConfig *config.Config `yaml:"config"`
+	BufferPeriod time.Duration  `yaml:"buffer_period"`
+	BufferCount  int            `yaml:"buffer_count"`
 }
 
+type preceiver struct {
+	startOnce sync.Once
+	recv      *promreceiver.Receiver
+	cfg       *Configuration
+}
+
+var _ receiver.MetricsReceiverSink = (*preceiver)(nil)
+
 // New creates a new prometheus.Receiver reference.
-func New(ms receiver.MetricsReceiverSink, opts ...Option) (*Receiver, error) {
-	if ms == nil {
-		return nil, errors.New("needs a non-nil receiver.MetricsReceiverSink")
-	}
-	pr := &Receiver{metricSink: ms}
-	for _, opt := range opts {
-		opt.WithReceiver(pr)
+func New(cfg *Configuration) (receiver.MetrcisReceiverSink, error) {
+	if cfg == nil || cfg.ScrapeConfig == nil {
+		return nil, errNilScrapeConfig
 	}
+	pr := &preceiver{cfg: cfg}
 	return pr, nil
 }
 
-const receiverName = "prometheus"
+var (
+	errAlreadyStarted         = errors.New("already started the Prometheus receiver")
+	errNilMetricsReceiverSink = errors.New("expecting a non-nil MetricsReceiverSink")
+	errNilScrapeConfig        = errors.New("expecting a non-nil ScrapeConfig")
+)
+
+func (pr *preceiver) StartMetricsReception(ctx context.Context, mrs receiver.MetricsReceiverSink) error {
+	var err = errAlreadyStarted
+	pr.startOnce.Do(func() {
+		if mrs == nil {
+			err = errNilMetricsReceiverSink
+			return
+		}
+
+		tms := &promMetricsReceiverToOpenCensusMetricsReceiver{mrs: mrs}
+		cfg := pr.cfg
+		pr.recv, err = promreceiver.ReceiverFromConfig(
+			context.Background(),
+			tms,
+			cfg.ScrapeConfig,
+			promreceiver.WithBufferPeriod(cfg.BufferPeriod),
+			promreceiver.WithBufferCount(cfg.BufferCount))
+	})
+	return err
```

Comment:
I believe this will always return this error (even when StartMetricsReception was called for the first time)

Reply:
No it won't. startOnce is only ever invoked once and when invoked on line 69, we set it
<img width="551" alt="screen shot 2019-02-06 at 6 51 50 pm" src="https://user-images.githubusercontent.com/4898263/52385761-4ecf6900-2a40-11e9-801e-b7b17189cffc.png">

Reply:
Ah idk how i missed that.... this LGTM!